### DISCUSSION TITLE
Question regarding creating general/table agnostic methods

### DISCUSSION THREAD

**Discussion by piperun** [simolus3/drift#3143] [Q&A] [answered]
I'm trying to write my own `connectOrCreate()` function that exists in Prisma, the issue I'm running into is that there doesn't seem to be a direct way of creating a general/generic/table agnostic solution from what I've scoured the documentation, so I'm wondering if there is one and I happen to miss it or am I out of luck?

This is somewhat what I want to be able to accomplish:
```dart
// database.dart
class AppDatabase extends _$AppDatabase {
  Future<void> connectOrCreate<T extends Table, D extends DataClass>(Insertable<D>  id, Insertable<D>  item) {
  // implementation...
}
```

```dart
// main.dart
void createData(String id, DbModel item) {
database.model.connectOrCreate(id, item);
}
```
Since right now I have to include a table inside the method like this:
```dart
// database
connectOrCreate<T extends Table, D extends DataClass>({
required TableInfo<T, D> table, required Insertable<D> id, required Isertable<D> item,
}) {
// Implementation...
}
// ---
// main.dart
void createData(String id, DbModel item) {
database.model.connectOrCreate(model, id, item);
// or
database.connectOrCreate(model, id, item);
}

```

**Comment by simolus3** [comment]
Using `TableInfo` to create table-agnostic queries is the way to go.

> This is somewhat what I want to be able to accomplish:

If `database.model.connectOrCreate(id, item);` is that you would like this to look like, can't you express this as an extension on `TableInfo`? E.g.

```dart
extension ConnectOrCreate<T extends Table, Row> on TableInfo<T, Row> {
  Future<void> connectOrCreate(Insertable<Row> id, Insertable<Row> item) {
    // As an example for a simplified implementation.
    return this.attachedDatabase.into(this).insert(item);
  }
}
```

Then you could call this with `database.model.connectOrCreate(id, item)`.

**Reply by piperun** [reply]
Wow! Thanks for the fast reply, I'll try out your suggestion!

**Reply by piperun** [reply]
The specific example you gave works without any issue, however when trying with selects, specifically trying to using `.where()` to filter, since the parameter within the arrow function will be of type T which means if I try and do: `where((tableItem) => tableItem.id.equals(id))` I'll get:
```
The getter 'id' isn't defined for the type 'T'.
Try importing the library that defines 'id', correcting the name to the name of an existing getter, or defining a getter or field named 'id'.
```

**Reply by simolus3** [reply, marked as answer]
You can use [schema introspection](https://drift.simonbinder.eu/docs/dart-api/schema_inspection/) to work around these issues. That's necessarily a bit unsafe, but there's no way to represent "extensions on tables having a column named id" in Darts type system, so that's the best we can do.

**Reply by piperun** [reply]
Understood, again thank you very much for the help!
